pkg/gen/dotnet-templates/Utilities.cs

```diff
@@ -1,26 +1,145 @@
-// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
+// Copyright 2016-2020, Pulumi Corporation
+// *** WARNING: this file was generated by the Pulumi Kubernetes codegen tool. ***
 // *** Do not edit by hand unless you're certain you know what you are doing! ***
 
 using System;
+using System.Diagnostics;
 using System.IO;
+using System.Linq;
 using System.Reflection;
-using Pulumi;
+using System.Text;
 
 namespace Pulumi.Kubernetes
 {
-    static class Utilities
+    internal static class Utilities
     {
-        private readonly static string version;
-        public static string Version => version;
+        public static string Version { get; }
 
         static Utilities()
         {
             var assembly = typeof(Utilities).GetTypeInfo().Assembly;
-            using (var stream = assembly.GetManifestResourceStream("Pulumi.Kubernetes.version.txt"))
-            using (var reader = new StreamReader(stream))
+            using var stream = assembly.GetManifestResourceStream("Pulumi.Kubernetes.version.txt");
+            if (stream == null)
+                throw new Exception("Manifest file 'version.txt' not found");
+            using var reader = new StreamReader(stream);
+            Version = reader.ReadToEnd().Trim();
+        }
+
+        public static InvokeOptions WithVersion(this InvokeOptions? options)
+        {
+            if (options?.Version != null)
+            {
+                return options;
+            }
+            return new InvokeOptions
+            {
+                Parent = options?.Parent,
```

Comment:
Hmm... This is copied from other providers which come from tfgen, e.g. https://github.com/pulumi/pulumi-azure/blob/master/sdk/dotnet/Utilities.cs#L49

Reply:
Asking differently, what is the test scenario which I could run to understand the effect of this?

Reply:
We were seeing hangs/crashes on preview if a Provider was specified for the resource when those invoke options were set.

Specifically, the invoke would always return an empty result structure. It may be a bug with the invoke implementation in the engine since I saw similar behavior in both NodeJS and Python.

Reply:
I double-checked and I still believe this is fine here:
- This method in `Utilities` simply copies the provider/parent from the options that are passed in as arguments + sets the version
- The options come from the invoke method, `YamlDecode` in our case
- We never call `YamlDecode` with explicit provider/parent
- `YamlDecode` is internal, so no outside users would call it differently